feat: add pytest comparison fixture

gohil-jay opened the pull request:
![image](https://user-images.githubusercontent.com/59703162/176820245-1bd81cc5-c837-46bd-8dd8-a73f36e386ca.png)

henryiii commented on src/boost_histogram/_internal/hist.py:
```diff
@@ -337,6 +338,24 @@ def ndim(self) -> int:
+                    return "The histogram contents : \n" + "\033[91m" + str(self.variances()) + "\033[0m" + "\nand\n" + "\033[91m" + str(hist2.variances()) + "\033[0m" + "\nare not equal."
+            else:
+                return "The histogram contents : \n" + "\033[91m" + str(self.view()) + "\033[0m" + "\nand\n" + "\033[91m" + str(hist2.view()) + "\033[0m" + "\nare not equal."
+            else:
```
No else after return, please! This would look much cleaner. Also, the final else block is misaligned. (CC @jpivarski this is an example of else after return. :)

I would then add "not" in front of each if, and place the return immediately after. Now you'll have a non-indented linear chain of checks, rather than this sideways pyramid. :)

gohil-jay replied:
Alright makes sense! Will make the changes :)

henryiii replied:
FYI, what I described is called "guard style". You check for the failure and immediately break the control flow (usually via `return`); everything after that is now guarded from that condition. It keeps logic simpler and reduces indentation.

Ruby actually has syntax for it, `return <value> if <condition>` can be written on one line, and the inverse is `return <value> unless <condition>`. Python never got custom syntax, since `if <condition>: return <value>` isn't actually any longer.

gohil-jay replied:
Interesting... Thanks for the insight!

henryiii commented on src/boost_histogram/_internal/hist.py:
```diff
@@ -337,6 +338,24 @@ def ndim(self) -> int:
+                        if (list(self.axes)==list(hist2.axes)):
+                            return ""
+                        else:
+                            return "The axes [" + "\033[91m" + str(list(self.axes)) + " and " + str(list(hist2.axes)) + "\033[0m" + "] are not equal."
```
```suggestion
                            return f"The axes [\033[91m{list(self.axes)} and {list(hist2.axes)}\033[0m] are not equal."
```

F-strings were made for this. :) Also I'd pull out the color formatting into a variable, like `red` and `reset`. Actually, there are lots of concerns with color formatting (needs to be disablable and such), at least rich can auto-format this, so probably leave the color formatting off for the moment.

henryiii replied:
I bet if you do `import rich; rich.print(h.compare(h2))` this would still be colorful. ;)

gohil-jay replied:
Yes I was going to do something similar (with termcolor), but then though I'd do it without importing a library..

gohil-jay commented:
@henryiii, pre-commit throws few errors... Any feedback/suggestion?

henryiii commented:
It's complaining because you are using the result of re.search without checking to see if it has a result first. I think the best fix, though, is to use `._storage_type`, which contains the storage type of the histogram, rather than trying to parse it out of the repr.

gohil-jay commented:
I've been looking for something like `_storage_type` since the beginning (couldn't find it on docs or book [here](https://boost-histogram.readthedocs.io/_/downloads/en/latest/pdf/))! But alright now it should work :)

gohil-jay commented:
Seems alright now (?)

henryiii commented:
Note: the goal is to support `pytest_assertrepr_compare` https://docs.pytest.org/en/latest/how-to/assert.html#defining-your-own-explanation-for-failed-assertions

henryiii committed:
feat: add a pytest plugin for histogram comparison

Assisted-by: ClaudeCode:claude-fable-5
Co-authored-by: Jay Gohil <jay.gohil.info@gmail.com>
Signed-off-by: Henry Schreiner <henryfs@princeton.edu>

henryiii committed:
test: skip plugin auto-load test in CMake-only builds

The end-to-end pytester test relies on the pytest11 entry point being
registered and on boost_histogram being importable in an isolated
subprocess. Neither holds in a CMake-only / in-place build (no pip
install), so guard it with skipif on the entry point's presence. The
direct unit tests still cover the diff output in all modes.

Assisted-by: ClaudeCode:claude-opus-4.8
